main/src/ca/uwaterloo/flix/language/phase/Optimizer.scala

```diff
@@ -41,14 +40,255 @@ object Optimizer extends Phase[SimplifiedAst.Root, SimplifiedAst.Root] {
     * Returns an optimized version of the given AST `root`.
     */
   def run(root: SimplifiedAst.Root)(implicit flix: Flix): Validation[SimplifiedAst.Root, CompilationError] = {
-    // Print the ast if debugging is enabled.
-    if (flix.options.debug) {
-      println(PrettyPrinter.Simplified.fmtRoot(root).fmt(TerminalContext.AnsiTerminal))
+
+    /**
+      * Returns `true` if and only if `enum` is a singleton-valued enum.
+      */
+    def isSingleCaseEnum(enum: Symbol.EnumSym): Boolean = root.enums.get(enum) match {
+      case None => false
+      case Some(defn) => defn.cases.size <= 1
     }
 
-    // TODO: Implement.
+    def toExp(b0: Boolean): Expression = if (b0) Expression.True else Expression.False
 
-    root.toSuccess
-  }
+    def optimizeBinaryExpression(e0: Expression.Binary): Expression = e0 match {
+      case Expression.Binary(op, exp1, exp2, tpe, loc) =>
+        val e1 = optimize(exp1)
+        val e2 = optimize(exp2)
+        (op, e1, e2) match {
+          // Elimination of run-time tag checks of singleton-valued enums.
+          case (BinaryOperator.Equal, Expression.Tag(sym, tag, exp, tp, lc), _) =>
```

Comment:
Let us not mess with equality, because it is hard to get right, and also because it is uncommon to do equality comparisons on enums. The real meat is in the `Is`, `Tag` and `Untag`.